### PR TITLE
Genesis chainconfig cleanup

### PR DESCRIPTION
**Pull request by ceyonur**
## Why this should be merged

1- This adds a new upgrade config `networkUpgradeOverrides` to override any avalanche network upgrade. With this upgrade config and `skip-upgrade-check` flag set to true, one can modify the timestamp of activated upgrade. This will come handy if a subnet has missed the upgrade (failed to update nodes) and as a result their chain has become unsyncable with newer versions. In this case `networkUpgradeOverrides` can be used to set a later time for upgrades. I.e:
`upgrade.json`:
```
{
  "networkUpgradeOverrides": {
    "durangoTimestamp": 1710415083
  }
}
```
2- Default Network Upgrades are introduced. `nil` and `blank` values in network upgrades (eth forks + avalanche upgrades) will now default to values. Meaning that it won't be possible to deactivate an existing upgrade via `nil` values. 

3- This PR also removes `OptionalNetworkUpgrades` since now we can modify all NetworkUpgrades.

## How this was tested

Unit tests + local tests

## How is this documented

it will be documented in docs repo.

### CODE REVIEW

File: params/config.go [hunk 1]
```diff
@@ -263,6 +257,42 @@ func (c *ChainConfig) Description() string {
 	return banner
 }
 
+func (c *ChainConfig) SetNetworkUpgradeDefaults() {
```
Comment (ceyonur) [inline params/config.go:260]: This will remove the capability of giving `nil` values for upgrades/forks to disable them. I don't think the current codebase supports deactivating any of those, but if we want to maintain the "full" backward-compatibility we can expand the `*uint64/*bigInt` types to some struct with which we can deduct information about `nil` vs `value` vs `blank` in json files. This would require a bigger change in the code and would increase the diff between the upstream.

File: params/network_upgrades.go [hunk 2]
```diff
@@ -4,63 +4,98 @@
...
 		return newTimestampCompatError("Durango fork block timestamp", m.DurangoTimestamp, newcfg.DurangoTimestamp)
 	}
-	if isForkTimestampIncompatible(m.CancunTime, newcfg.CancunTime, time) {
-		return newTimestampCompatError("Cancun fork block timestamp", m.CancunTime, m.CancunTime)
-	}
 	return nil
 }
 
-func (m *MandatoryNetworkUpgrades) mandatoryForkOrder() []fork {
+func (m *NetworkUpgrades) forkOrder() []fork {
```
Comment (darioush) [inline params/network_upgrades.go:40]: nit: should we rename receivers to `n`?

File: core/state_processor.go [hunk 3]
```diff
@@ -197,7 +198,15 @@ func ApplyPrecompileActivations(c *params.ChainConfig, parentTimestamp *uint64,
...
-				log.Info("Activating new precompile", "name", module.ConfigKey, "config", activatingConfig)
+				var printIntf interface{}
+				marshalled, err := json.Marshal(activatingConfig)
+				if err == nil {
+					printIntf = string(marshalled)
+				} else {
+					printIntf = activatingConfig
+				}
+
+				log.Info("Activating new precompile", "name", module.ConfigKey, "config", printIntf)
```
Comment (darioush) [inline core/state_processor.go:209]: should we require a `.String()` on the config instead?

File: plugin/evm/vm_test.go [hunk 4]
```diff
@@ -2819,7 +2819,7 @@ func TestRewardManagerPrecompileSetRewardAddress(t *testing.T) {
 	data, err := rewardmanager.PackSetRewardAddress(testAddr)
 	require.NoError(t, err)
 
-	gas := 21000 + 240 + rewardmanager.SetRewardAddressGasCost // 21000 for tx, 240 for tx data
+	gas := 21000 + 240 + rewardmanager.SetRewardAddressGasCost + rewardmanager.RewardAddressChangedEventGasCost // 21000 for tx, 240 for tx data
```
Comment (darioush) [inline plugin/evm/vm_test.go:2822]: why are these changed?

Comment (ceyonur) [reply plugin/evm/vm_test.go:2822]: because VMs will now have non-nil upgrades (if default is defined). This was running with only `subnetEVM` upgrade (Durango was nil in chainconfig) but since now we apply defaults for nils, this test is upgraded to Durango fork. In Durango precompiles have increased costs (because of events)

File: core/state_processor.go (hunk 3 above)
Comment (ceyonur) [reply core/state_processor.go:209]: I thought about it but that would require a change to all precompiles and i don't if we there is any gain with custom string configs vs plain marshalling them to json.

**Review by darioush — APPROVED** [review]
